resource/resource.go

```diff
@@ -168,46 +151,145 @@ func (r *Resource) Out(inputDirectory string, source oc.Source, params oc.Params
+			if err != nil {
+				return nil, nil, err
+			}
+
+			err = copySecretToVault(r.client, p.Prefix, secretMap.Dest, secretFile, secret)
+			if err != nil {
+				return nil, nil, err
 			}
-			secert := sv.NewSecret()
-			err = secert.UnmarshalJSON(bytes)
+
+		}
+	}
+	// Both `version` and `metadata` may be empty. In this case, we are returning
+	// `version` retrieved from the file created by `In`, while `metadata` is empty.
+	metadata := oc.Metadata{}
+	return nil, metadata, nil
+}
+
+func listFilesUnder(rootDir string) ([]string, error) {
+	ret := []string{}
+	err := filepath.Walk(rootDir,
+		func(path string, info os.FileInfo, err error) error {
 			if err != nil {
 				return err
 			}
+			if info.IsDir() {
+				return nil
+			}
 			secretPath, err := filepath.Rel(rootDir, path)
 			if err != nil {
 				return err
 			}
-			return r.client.Write(filepath.Join(p.Prefix, secretPath), secert)
-		})
+			ret = append(ret, secretPath)
+			return nil
+		},
+	)
+	return ret, err
+}
+
+func createSecret(rootDir string, secretFile string) (*sv.Secret, error) {
+	srcFile, err := os.Open(filepath.Join(rootDir, secretFile))
 	if err != nil {
-		return nil, nil, err
+		return nil, fmt.Errorf("Error reading source file '%s': %s", secretFile, err)
+	}
+	jDec := json.NewDecoder(srcFile)
+	secret := sv.NewSecret()
+	err = jDec.Decode(&secret)
+	if err != nil {
+		return nil, err
 	}
+	return secret, nil
+}
 
-	// Both `version` and `metadata` may be empty. In this case, we are returning
-	// `version` retrieved from the file created by `In`, while `metadata` is empty.
-	metadata := oc.Metadata{}
-	return nil, metadata, nil
+func validate(secret *sv.Secret, keys []interface{}) error {
+	finalKeys := getFinalKeys(keys)
+	ok := true
+	message := "Specified keys not found:"
+	for key := range finalKeys {
+		if !secret.Has(key) {
+			ok = false
+			message += fmt.Sprintf(" '%s'", key)
+		}
+	}
+	if !ok {
+		return fmt.Errorf(message)
+	}
+	return nil
+}
+
+func filterAndRenameKeys(secret *sv.Secret, keys []interface{}) error {
+	if len(keys) == 0 {
+		return nil
+	}
+	finalKeys := getFinalKeys(keys)
+	for _, currentKey := range secret.Keys() {
+		finalKey, isFinalKey := finalKeys[currentKey]
+		if isFinalKey {
+			value := secret.Get(currentKey)
+			err := secret.Set(finalKey, value, false)
+			if err != nil {
+				return err
+			}
+		}
+		if (!isFinalKey) || (finalKey != currentKey) {
+			secret.Delete(currentKey)
+		}
+	}
+	return nil
+}
+
+func copySecretToVault(client *sv.Vault, prefix string, dest string, secretPath string, secret *sv.Secret) error {
+	finalDest := filepath.Join(prefix, dest, secretPath)
+	return client.Write(finalDest, secret)
+}
+
+func getFinalKeys(keys []interface{}) map[string]string {
+	finalKeys := map[string]string{}
+	for _, key := range keys {
+		switch v := key.(type) {
+		case string:
+			keyString := fmt.Sprintf("%v", key)
+			finalKeys[keyString] = keyString
+		case map[string]string:
```

Comment:
I would add a check after this line to assert that the length of the map is only one, because it doesn't really make sense to have more than one key in the map for our schema.